src/cpp/fastdds/builtin/type_lookup_service/TypeLookupRequestListener.cpp

```diff
@@ -40,104 +36,406 @@ namespace fastdds {
+            requests_queue_.pop();
+        }
+    }
+}
 
-                TypeLookup_Reply* reply = static_cast<TypeLookup_Reply*>(typelookup_manager_->reply_type_.createData());
-                TypeLookup_getTypes_Result result;
-                result.result(out);
-                reply->return_value().getType(result);
-                reply->header().relatedRequestId() = request.header().requestId();
+void TypeLookupRequestListener::check_get_types_request(
+        SampleIdentity request_id,
+        const TypeLookup_getTypes_In& request)
+{
+    // Always Sends EK_COMPLETE
+    // TODO: Add a property to the participant to configure this behavior. Allowing it to respond with EK_MINIMAL when possible.
+    TypeLookup_getTypes_Out out;
+    ReturnCode_t type_result = RETCODE_ERROR;
+    xtypes::TypeObject obj;
+    xtypes::TypeIdentifier complete_id;
+    xtypes::TypeIdentifier minimal_id;
+    // Iterate through requested type_ids
+    for (const xtypes::TypeIdentifier& type_id : request.type_ids())
+    {
+        // If TypeIdentifier is EK_MINIMAL add complete_to_minimal to answer
+        if (type_id._d() == xtypes::EK_MINIMAL)
+        {
+            minimal_id = type_id;
+            // Get complete TypeIdentifier from registry
+            complete_id = fastrtps::rtps::RTPSDomainImpl::get_instance()->type_object_registry_observer().
+                            get_complementary_type_identifier(minimal_id);
 
-                typelookup_manager_->send_reply(*reply);
-                typelookup_manager_->reply_type_.deleteData(reply);
+            xtypes::TypeIdentifierPair id_pair;
+            id_pair.type_identifier1(complete_id);
+            id_pair.type_identifier2(minimal_id);
+            // Add the id pair to the result
+            out.complete_to_minimal().push_back(std::move(id_pair));
+        }
+        else
+        {
+            complete_id = type_id;
+        }
 
-                break;
+        type_result = fastrtps::rtps::RTPSDomainImpl::get_instance()->type_object_registry_observer().
+                        get_type_object(complete_id, obj);
+        if (RETCODE_OK != type_result)
+        {
+            // If any object is unknown, abort and answer with exception
+            break;
+        }
+
+        xtypes::TypeIdentifierTypeObjectPair id_obj_pair;
+        id_obj_pair.type_identifier(complete_id);
+        id_obj_pair.type_object(obj);
+        // Add the id/obj pair to the result
+        out.types().push_back(std::move(id_obj_pair));
+    }
+
+    if (RETCODE_OK == type_result)
+    {
+        // Prepare and send the reply
+        answer_request(request_id, rpc::RemoteExceptionCode_t::REMOTE_EX_OK, out);
+    }
+    else
+    {
+        // If any of the types is not found, log error
+        EPROSIMA_LOG_WARNING(TYPELOOKUP_SERVICE_REQUEST_LISTENER, "Error getting type.");
+        answer_request(request_id, rpc::RemoteExceptionCode_t::REMOTE_EX_INVALID_ARGUMENT);
+    }
```

Comment:
I think we should distinguish between the return error RETCODE_NO_DATA and the RETCODE_PRECONDITION_NOT_MET. The first one is returned if the registry does not have information about the type being queried, REMOTE_EX_UNKNOWN EXCEPTION should be returned because the received argument is not per se invalid, only unknown. The second one, receiving an invalid argument, should return the REMOTE_EX_INVALID_ARGUMENT.

Reply:
added better error check in [d054d39](https://github.com/eProsima/Fast-DDS/pull/4240/commits/d054d39484489f3890638c8cfe4fe4c023139c74)